tutorbackup/templates/backup/build/backup/restore_services.py

```diff
@@ -0,0 +1,189 @@
+                   stdout=sys.stdout,
+                   stderr=sys.stderr)
+
+    logger.info("MySQL restored.")
+
+
+def restore_mongodb():
+    host = ENV['MONGODB_HOST']
+    port = ENV['MONGODB_PORT']
+    dump_dir = MONGODB_DUMPDIR
+
+    logger.info(
+        f"Restoring MongoDB databases on {host}:{port} from {dump_dir}")
+    cmd = ("mongorestore "
+           "--stopOnError --drop "
+           f"--host={host} --port={port} "
+           f"{dump_dir}"
+           )
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    logger.info("MongoDB restored.")
+
+
+def restore_caddy():
+    dump_dir = CADDY_DUMPDIR
+    caddy_dir = 'caddy'
+    logger.info(f"Copying Caddy data from {dump_dir}")
+    shutil.copytree(dump_dir, caddy_dir)
+
+    total_size = get_size(caddy_dir)
+    logger.info(f"Complete. {caddy_dir} total size {total_size} bytes.")
+
+
+def extract():
+    tar_file = TARFILE
+    out_dir = DUMP_DIRECTORY
+
+    logger.info(f"Extracting archive {tar_file} to {out_dir}")
+    with tarfile.open(tar_file, "r:xz") as tar:
+        tar.extractall()
+
+    size = get_size(out_dir)
+    logger.info(f"Complete. {out_dir} is {size} bytes.")
+
+
+def download_from_s3(version_id=None):
+    from s3_client import S3_CLIENT, IntegrityError
+
+    bucket = ENV['S3_BUCKET_NAME']
+    file_name = TARFILE
+
+    logger.info(f"Downloading {file_name} from S3 bucket {bucket}")
+    try:
+        S3_CLIENT.download_file(
+            bucket,
+            os.path.basename(file_name),
+            file_name,
+            ExtraArgs={'VersionId': version_id} if version_id else None,
+        )
+
+        logger.info("Checking downloaded file's integrity ...")
+        obj_metadata = S3_CLIENT.head_object(
+            Bucket=bucket,
+            Key=os.path.basename(file_name),
+        )
+
+        received_size = obj_metadata['ContentLength']
+        size = os.path.getsize(file_name)
+        size_correct = received_size == size
+
+        received_version_id = obj_metadata['VersionId']
+        if version_id:
+            version_id_correct = version_id == received_version_id
+        else:
+            version_id_correct = True
+
+        received_checksum = obj_metadata['ETag'].strip('"')
+
+        if size_correct and version_id_correct:
+            logger.info(f"File integrity verified.\n"
+                        f"Version ID: '{received_version_id}'\n"
+                        f"Size: {size} bytes\n"
+                        f"Checksum: '{received_checksum}'")
+        else:
+            os.remove(file_name)
+            raise IntegrityError(
+                f"File integrity could not be verified. "
+                f"Deleted the downloaded file with "
+                f"VersionId='{received_version_id}'.")
```

Comment:
Minor style issue: I tend to write this as

```python
raise IntegrityError(
    "File integrity could not be verified. "
    "Deleted the downloaded file with "
    f"VersionId='{received_version_id}'.")
```

As in, only add the `f` prefix to strings where something is actually being replaced.

Reply:
Thanks. I missed that one in both scripts. Will fix.